docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+  `DisruptionBudgetSpec`. In this approach, we assume a disruption budget controller
+  keeps the `DisruptionBudgetStatus` up-to-date by observing all pod deletions and
+  creations in the cluster, so that an approved disruption is quickly reflected in the
+  `DisruptionBudgetStatus`. Of course this approach does allow a race in which one or
+  more additional disruptions could be approved before the first one is reflected in the
+  `DisruptionBudgetStatus`.
+    * Thus a subsequent implementation will have the API server explicitly debit the
+  `DisruptionBudgetStatus` when it accepts an `/evict`. (There still needs to be a
+  controller, to keep the shard strength status up-to-date when replacement pods are
+  created after an eviction; the controller may also be necessary for the rate status
+  depending on how rate is represented, e.g. adding tokens to a bucket at a fixed rate.)
+  Once etcd support multi-object transactions (etcd v3), the debit and pod deletion will
+  be placed in the same transaction.
+    * Note: For the purposes of disruption budget, a pod is considered to be disrupted as soon as its
+  graceful termination period starts (so when we say "delete" here we do not mean
+  "deleted from etcd" but rather "graceful termination period has started").
+  * It will allow clients to communicate additional parameters when they wish to delete a
+  pod. (In the absence of the `/evict` subresource, we would have to create a pod-specific
+  type analogous to `api.DeleteOptions`.)
+
+We will make `kubectl delete pod` use `/evict` by default, and require a command-line
+flag to delete the pod directly.
+
+We will add to `NodeStatus` a bounded-sized list of signatures of pods that should avoid
+that node (provisionally called `PreferAvoidPods`). One of the pieces of information
+specified in the `/evict` subresource is whether the eviction should add the evicted
+pod's signature to the corresponding node's `PreferAvoidPods`. Initially the pod
+signature will be a
+[controllerRef](https://github.com/kubernetes/kubernetes/issues/14961#issuecomment-183431648),
+i.e. a reference to the pod's controller. Controllers are responsible for garbage
+collecting, after some period of time, `PreferAvoidPods` that point to them, but the API
+server will also enforce a bounded size on the list. All schedulers will have a
+highest-weighted priority function that gives a node the worst priority if the pod it is
+scheduling appears in that node's `PreferAvoidPods` list. Thus appearing in
+`PreferAvoidPods` is similar to
+[RequiredDuringScheduling node anti-affinity](../../docs/user-guide/node-selection/README.md)
+but it takes precedence over all other priority criteria and is not explicitly listed in
+the `NodeAffinity` of the pod.
+
+`PreferAvoidPods` is useful for the "moving a running pod off of a node from which it is
+receiving poor service" use case, as it reduces the chance that the replacement pod will
```

Comment:
Not yet. I anticipate that the people who implement the rescheduler/experiment with rescheduler policies will write a doc covering that question (for the policies they plan to implement) before implementing. In this doc I've just sketched out some general categories; defining the exact criteria you'd use for triggering an eviction is worthy of (a) separate doc(s).